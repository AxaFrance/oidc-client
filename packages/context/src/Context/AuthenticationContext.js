import React from 'react';
import PropTypes from 'prop-types';
import { AuthenticationContext } from './AuthenticationContextCreator';
import { OidcRoutes } from '../Routes';

const propTypes = {
  notAuthentified: PropTypes.node,
  notAuthorized: PropTypes.node,
  isLoading: PropTypes.bool.isRequired,
  isEnabled: PropTypes.bool,
  // eslint-disable-next-line
  oidcUser: PropTypes.object, //TODO : rÃ©cuperer le proptypes depuis OIDC client ?
  error: PropTypes.string.isRequired,
  login: PropTypes.func.isRequired,
  logout: PropTypes.func.isRequired,
};

const defaultProps = {
  notAuthentified: null,
  notAuthorized: null,
  isEnabled: true,
};

const AuthenticationProviderComponent = ({
  isLoading,
  isEnabled,
  oidcUser,
  error,
  login,
  logout,
  notAuthentified,
  notAuthorized,
  children,
}) => (
  <AuthenticationContext.Provider
    value={{
      isLoading,
      oidcUser,
      error,
      login,
      logout,
      isEnabled,
    }}
  >
    <OidcRoutes notAuthentified={notAuthentified} notAuthorized={notAuthorized}>
      {children}
    </OidcRoutes>
  </AuthenticationContext.Provider>
);

AuthenticationProviderComponent.propTypes = propTypes;
AuthenticationProviderComponent.defaultProps = defaultProps;

export default AuthenticationProviderComponent;
