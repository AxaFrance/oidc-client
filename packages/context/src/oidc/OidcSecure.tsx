import React, {useEffect, PropsWithChildren, FC} from 'react';

import Oidc from "./vanilla/oidc";

export type OidcSecureProps = {
    callbackPath?:string;
    configurationName?: string;
};

export const OidcSecure: FC<PropsWithChildren<OidcSecureProps>> = ({children, callbackPath=null, configurationName="default"}) => {
    const getOidc =  Oidc.get;
    const oidc = getOidc(configurationName);
    useEffect(() => {
        if(!oidc.tokens){
            oidc.loginAsync(callbackPath);
        }
        return () => {
        }
    }, [configurationName])

    if(!oidc.tokens){
      return null;
    }
    return <>{children}</>
}

export const withOidcSecure = (
  WrappedComponent, callbackPath=null
) => (props) => {
  return <OidcSecure callbackPath={callbackPath}><WrappedComponent {...props} /></OidcSecure>;
};