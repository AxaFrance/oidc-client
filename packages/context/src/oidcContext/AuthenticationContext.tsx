import React, { ComponentType } from 'react';
import { User, UserManagerEvents, UserManager } from 'oidc-client';

export type oidcContext = {
  oidcUser: User | null;
  isEnabled: boolean;
  login: Function;
  logout: Function;
  events: UserManagerEvents;
  authenticating: ComponentType;
  isLoading: boolean;
  userManager: UserManager;
  error: string;
};

export const AuthenticationContext = React.createContext<oidcContext>(null);

export const useReactOidc = () => {
  const { isEnabled, login, logout, oidcUser, events } = React.useContext(AuthenticationContext);
  return { isEnabled, login, logout, oidcUser, events };
};
