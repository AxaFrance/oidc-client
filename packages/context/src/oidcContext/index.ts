export { default as AuthenticationProvider } from './AuthenticationContext.provider';

export { AuthenticationContext, useReactOidc } from './AuthenticationContext';
