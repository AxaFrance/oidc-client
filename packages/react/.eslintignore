build
bin
node_modules
dist
**/*.spec.ts
**/*.spec.tsx
**/*.test.ts
**/*.test.tsx
**/*.stories.tsx
./src/setupTests.ts
setupTests.js
.eslintrc.js
config
jest.config.js
App.specold.tsx
