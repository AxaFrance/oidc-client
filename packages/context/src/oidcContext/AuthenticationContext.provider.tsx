import React, { useEffect, useCallback, ComponentType, PropsWithChildren } from 'react';
import PropTypes from 'prop-types';
import { User, Logger, UserManagerSettings, UserManagerEvents } from 'oidc-client';
import {
  withRouter,
  authenticationService,
  setLogger,
  OidcRoutes,
  configurationPropTypes,
  configurationDefaultProps,
  ReactOidcHistory,
  UserStoreType,
  oidcLog,
  authenticateUser,
  logoutUser,
} from '@axa-fr/react-oidc-core';

import { Callback } from '../Callback';

import withServices from '../withServices';
import { AuthenticationContext } from './AuthenticationContext';
import { useAuthenticationContextState, useOidcEvents, CustomEvents } from './AuthenticationContext.hooks';

type AuthenticationProviderIntProps = PropsWithChildren<{
  location: Location;
  history: ReactOidcHistory;
  loggerLevel: number;
  logger: Logger;
  notAuthenticated: ComponentType;
  notAuthorized: ComponentType;
  authenticating: ComponentType;
  callbackComponentOverride: ComponentType;
  sessionLostComponent: ComponentType;
  UserStore: UserStoreType;
  isEnabled?: boolean;
  configuration: UserManagerSettings;
  authenticationServiceInt: typeof authenticationService;
  CallbackInt: typeof Callback;
  setLoggerInt: typeof setLogger;
  OidcRoutesInt: typeof OidcRoutes;
  oidcLogInt: typeof oidcLog;
  authenticateUserInt: typeof authenticateUser;
  logoutUserInt: typeof logoutUser;
  customEvents: CustomEvents;
}>;

const propTypes = {
  notAuthenticated: PropTypes.elementType,
  notAuthorized: PropTypes.elementType,
  authenticating: PropTypes.elementType,
  callbackComponentOverride: PropTypes.elementType,
  sessionLostComponent: PropTypes.elementType,
  configuration: configurationPropTypes,
  isEnabled: PropTypes.bool,
  loggerLevel: PropTypes.number,
  logger: PropTypes.shape({
    info: PropTypes.func.isRequired,
    warn: PropTypes.func.isRequired,
    error: PropTypes.func.isRequired,
    debug: PropTypes.func.isRequired,
  }),
  UserStore: PropTypes.func,
  customEvents: PropTypes.shape({
    onUserLoaded: PropTypes.func,
    onUserUnloaded: PropTypes.func,
    onSilentRenewError: PropTypes.func,
    onUserSignedOut: PropTypes.func,
    onUserSessionChanged: PropTypes.func,
    onAccessTokenExpiring: PropTypes.func,
    onAccessTokenExpired: PropTypes.func,
  }),
};

const defaultProps: Partial<AuthenticationProviderIntProps> = {
  notAuthenticated: null,
  notAuthorized: null,
  authenticating: null,
  callbackComponentOverride: null,
  sessionLostComponent: null,
  isEnabled: true,
  loggerLevel: 0,
  logger: console,
  configuration: configurationDefaultProps,
  customEvents: null,
};

export const withComponentOverrideProps = (Component: ComponentType, customCallback: ComponentType) => (props: PropsWithChildren<any>) => (
  <Component callbackComponentOverride={customCallback} {...props} />
);

export const AuthenticationProviderInt = ({
  location,
  history,
  configuration,
  isEnabled,
  UserStore,
  loggerLevel,
  logger,
  sessionLostComponent,
  authenticating,
  notAuthenticated,
  notAuthorized,
  callbackComponentOverride,
  children,
  customEvents,
  // Injected
  authenticationServiceInt,
  CallbackInt,
  setLoggerInt,
  OidcRoutesInt,
  oidcLogInt,
  authenticateUserInt,
  logoutUserInt,
}: AuthenticationProviderIntProps) => {
  const userManager = authenticationServiceInt(configuration, UserStore);
  const { oidcState, loadUser, onError, onLoading, unloadUser, onLogout } = useAuthenticationContextState(userManager);
  const oidcFunctions = { loadUser, onError, onLoading, unloadUser, onLogout };
  const { addOidcEvents, removeOidcEvents } = useOidcEvents(oidcLogInt, userManager, oidcFunctions, customEvents);

  useEffect(() => {
    onLoading();
    setLoggerInt(loggerLevel, logger);
    addOidcEvents();
    let mount = true;
    userManager.getUser().then((user: User | null) => {
      if (mount) {
        loadUser(user);
      }
    });
    return () => {
      removeOidcEvents();
      mount = false;
    };
  }, [addOidcEvents, loadUser, logger, loggerLevel, onLoading, removeOidcEvents, setLoggerInt, userManager, customEvents]);

  const CallbackComponent = React.useMemo(
    () => (callbackComponentOverride ? withComponentOverrideProps(CallbackInt, callbackComponentOverride) : CallbackInt),
    [CallbackInt, callbackComponentOverride]
  );

  const loginCallback = useCallback(async () => {
    onLoading();
    oidcLogInt.info('Login requested');
    await authenticateUserInt(userManager, location, history)();
  }, [authenticateUserInt, history, location, oidcLogInt, onLoading, userManager]);

  const logoutCallback = useCallback(async () => {
    try {
      onLogout();
      await logoutUserInt(userManager);
      oidcLogInt.info('Logout successfull');
    } catch (error) {
      onError(error.message);
    }
  }, [logoutUserInt, oidcLogInt, onError, onLogout, userManager]);
  return (
    <AuthenticationContext.Provider
      value={{
        ...oidcState,
        authenticating,
        isEnabled,
        login: loginCallback,
        logout: logoutCallback,
        events: oidcState.userManager.events,
      }}
    >
      <OidcRoutesInt
        notAuthenticated={notAuthenticated}
        notAuthorized={notAuthorized}
        callbackComponent={CallbackComponent}
        sessionLost={sessionLostComponent}
        configuration={configuration}
      >
        {children}
      </OidcRoutesInt>
    </AuthenticationContext.Provider>
  );
};

type AuthenticationProviderProps = Omit<AuthenticationProviderIntProps,
  | 'authenticationServiceInt'
  | 'CallbackInt'
  | 'setLoggerInt'
  | 'OidcRoutesInt'
  | 'oidcLogInt'
  | 'authenticateUserInt'
  | 'logoutUserInt'
>

const AuthenticationProvider: ComponentType<Partial<AuthenticationProviderProps>> = withRouter(
  withServices(AuthenticationProviderInt, {
    CallbackInt: Callback,
    authenticationServiceInt: authenticationService,
    setLoggerInt: setLogger,
    OidcRoutesInt: OidcRoutes,
    oidcLogInt: oidcLog,
    authenticateUserInt: authenticateUser,
    logoutUserInt: logoutUser,
  })
);
// @ts-ignore
AuthenticationProvider.propTypes = propTypes;
AuthenticationProvider.defaultProps = defaultProps;
AuthenticationProvider.displayName = 'AuthenticationProvider';

export default AuthenticationProvider;
