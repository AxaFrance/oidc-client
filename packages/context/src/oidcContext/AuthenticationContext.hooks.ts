import { useReducer, Dispatch, useCallback } from 'react';
import { User, UserManager, Logger } from 'oidc-client';

// useAuthenticationContextState hook part

export interface OidcState {
  oidcUser: User | null;
  userManager: UserManager;
  isLoading: boolean;
  error: string;
}

export interface UseAuthenticationContextStateType {
  onError: Function;
  loadUser: Function;
  onLoading: Function;
  unloadUser: Function;
  oidcState: OidcState;
}

const ON_LOADING = 'ON_LOADING';
const ON_ERROR = 'ON_ERROR';
const ON_LOAD_USER = 'ON_LOAD_USER';
const ON_UNLOAD_USER = 'ON_UNLOAD_USER';

type OidcAction =
  | { type: 'ON_LOADING' }
  | { type: 'ON_ERROR'; message: string }
  | { type: 'ON_LOAD_USER'; user: User | null }
  | { type: 'ON_UNLOAD_USER' };

const getDefaultState = (userManagerInt: UserManager): OidcState => {
  return {
    oidcUser: null,
    userManager: userManagerInt,
    isLoading: false,
    error: '',
  };
};

const oidcReducer = (oidcState: OidcState, action: OidcAction): OidcState => {
  switch (action.type) {
    case ON_ERROR:
      return { ...oidcState, error: action.message, isLoading: false };
    case ON_LOADING:
      return { ...oidcState, isLoading: true };
    case ON_LOAD_USER:
      return { ...oidcState, oidcUser: action.user, isLoading: false };
    case ON_UNLOAD_USER:
      return { ...oidcState, oidcUser: null, isLoading: false };
    default:
      return oidcState;
  }
};

const onError = (dispatch: Dispatch<OidcAction>) => (message: string) => dispatch({ type: ON_ERROR, message });
const loadUser = (dispatch: Dispatch<OidcAction>) => (user: User | null) => dispatch({ type: ON_LOAD_USER, user });
const onLoading = (dispatch: Dispatch<OidcAction>) => () => dispatch({ type: ON_LOADING });
const unloadUser = (dispatch: Dispatch<OidcAction>) => () => dispatch({ type: ON_UNLOAD_USER });

export const useAuthenticationContextState = (userManagerInt: UserManager): UseAuthenticationContextStateType => {
  const defaultState = getDefaultState(userManagerInt);
  const [oidcState, dispatch] = useReducer(oidcReducer, defaultState);

  return {
    onError: useCallback(error => onError(dispatch)(error), []),
    loadUser: useCallback(user => loadUser(dispatch)(user), []),
    onLoading: useCallback(() => onLoading(dispatch)(), []),
    unloadUser: useCallback(() => unloadUser(dispatch)(), []),
    oidcState,
  };
};

// useOidcEvents Hook part

type OidcFunctions = Omit<UseAuthenticationContextStateType, 'oidcState'>;

export const onErrorEvent = (oidcLog: Logger, onErrorInt: Function) => (error: Error) => {
  oidcLog.error(`Error : ${error.message}`);
  onErrorInt(error.message);
};

export const onUserLoadedEvent = (oidcLog: Logger, loadUserInt: Function) => (user: User | null) => {
  oidcLog.info('User Loaded');
  loadUserInt(user);
};

export const onUserUnloadedEvent = (oidcLog: Logger, unloadUserInternal: Function) => () => {
  oidcLog.info('User unloaded');
  unloadUserInternal();
};

export const onAccessTokenExpiredEvent = (oidcLog: Logger, unloadUserInternal: Function, userManager: UserManager) => async () => {
  oidcLog.info('AccessToken Expired');
  unloadUserInternal();
  await userManager.signinSilent();
};

export const useOidcEvents = (oidcLog: Logger, userManager: UserManager, oidcFunctions: OidcFunctions) => {
  const addOidcEvents = useCallback(() => {
    userManager.events.addUserLoaded(onUserLoadedEvent(oidcLog, oidcFunctions.loadUser));
    userManager.events.addSilentRenewError(onErrorEvent(oidcLog, oidcFunctions.onError));
    userManager.events.addUserUnloaded(onUserUnloadedEvent(oidcLog, oidcFunctions.unloadUser));
    userManager.events.addUserSignedOut(onUserUnloadedEvent(oidcLog, oidcFunctions.unloadUser));
    userManager.events.addAccessTokenExpired(onAccessTokenExpiredEvent(oidcLog, oidcFunctions.unloadUser, userManager));
  }, [oidcFunctions.loadUser, oidcFunctions.onError, oidcFunctions.unloadUser, oidcLog, userManager]);

  const removeOidcEvents = useCallback(() => {
    userManager.events.removeUserLoaded(onUserLoadedEvent(oidcLog, oidcFunctions.loadUser));
    userManager.events.removeSilentRenewError(onErrorEvent(oidcLog, oidcFunctions.onError));
    userManager.events.removeUserUnloaded(onUserUnloadedEvent(oidcLog, oidcFunctions.unloadUser));
    userManager.events.removeUserSignedOut(onUserUnloadedEvent(oidcLog, oidcFunctions.unloadUser));
    userManager.events.removeAccessTokenExpired(onAccessTokenExpiredEvent(oidcLog, oidcFunctions.unloadUser, userManager));
  }, [oidcFunctions.loadUser, oidcFunctions.onError, oidcFunctions.unloadUser, oidcLog, userManager]);

  return { addOidcEvents, removeOidcEvents };
};
