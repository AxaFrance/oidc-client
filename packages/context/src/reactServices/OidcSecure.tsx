import React, { ComponentType, PropsWithChildren, useContext, useEffect, useMemo } from 'react';
import { UserManager } from 'oidc-client';
import {
  withRouter,
  Authenticating,
  getUserManager,
  authenticateUser,
  isRequireAuthentication,
  oidcLog,
  ReactOidcHistory,
} from '@axa-fr/react-oidc-core';

import withServices from '../withServices';
import { AuthenticationContext } from '../oidcContext';

type OidcComponentProps = PropsWithChildren<{
  location: Location;
  history: ReactOidcHistory;
  authenticateUserInternal: typeof authenticateUser;
  getUserManagerInternal: typeof getUserManager;
  isRequireAuthenticationInternal: typeof isRequireAuthentication;
  AuthenticatingInternal: typeof Authenticating;
  children?: ComponentType;
}>;

export const useOidcSecure = (
  authenticateUserInternal: typeof authenticateUser,
  userManager: UserManager,
  location: Location,
  history: ReactOidcHistory,
  oidcLogInternal: typeof oidcLog,
  AuthenticatingInternal: typeof Authenticating,
  isRequireAuthenticationInternal: typeof isRequireAuthentication,
  WrappedComponent: ComponentType
): ComponentType => {
  const { isEnabled, oidcUser, authenticating } = useContext(AuthenticationContext);
  useEffect(() => {
    oidcLogInternal.info('Protection : ', isEnabled);
    if (isEnabled) {
      oidcLogInternal.info('Protected component mounted');
      authenticateUserInternal(userManager, location, history)();
    }
    return () => {
      oidcLogInternal.info('Protected component unmounted');
    };
  }, [isEnabled, authenticateUserInternal, userManager, oidcLogInternal, location, history]);
  const requiredAuth = useMemo(() => isRequireAuthenticationInternal(oidcUser, false) && isEnabled, [
    isEnabled,
    isRequireAuthenticationInternal,
    oidcUser,
  ]);
  const AuthenticatingComponent: ComponentType = authenticating || AuthenticatingInternal;
  return requiredAuth ? AuthenticatingComponent : WrappedComponent;
};

export const OidcSecureWithInjectedFunctions = ({
  children,
  location,
  history,
  authenticateUserInternal,
  getUserManagerInternal,
  isRequireAuthenticationInternal,
  AuthenticatingInternal,
}: OidcComponentProps) => {
  const userManager = getUserManagerInternal();
  const WrappedComponent = () => <>{children}</>;
  const ReactOidcComponent = useOidcSecure(
    authenticateUserInternal,
    userManager,
    location,
    history,
    oidcLog,
    AuthenticatingInternal,
    isRequireAuthenticationInternal,
    WrappedComponent
  );

  return <ReactOidcComponent />;
};

export const withOidcSecureWithInjectedFunctions = (WrappedComponent: ComponentType) => ({
  location,
  history,
  authenticateUserInternal,
  getUserManagerInternal,
  isRequireAuthenticationInternal,
  AuthenticatingInternal,
  ...otherProps
}: OidcComponentProps) => {
  const userManager = getUserManagerInternal();
  const OverrideWrappedComponent = () => <WrappedComponent {...otherProps} />;

  const ReactOidcComponent = useOidcSecure(
    authenticateUserInternal,
    userManager,
    location,
    history,
    oidcLog,
    AuthenticatingInternal,
    isRequireAuthenticationInternal,
    OverrideWrappedComponent
  );

  return <ReactOidcComponent />;
};

export const withOidcSecure = (WrappedComponent: ComponentType) =>
  withRouter(
    withServices(withOidcSecureWithInjectedFunctions(WrappedComponent), {
      authenticateUserInternal: authenticateUser,
      getUserManagerInternal: getUserManager,
      isRequireAuthenticationInternal: isRequireAuthentication,
      AuthenticatingInternal: Authenticating,
    })
  );

const OidcSecure = withRouter(
  withServices(OidcSecureWithInjectedFunctions, {
    authenticateUserInternal: authenticateUser,
    getUserManagerInternal: getUserManager,
    isRequireAuthenticationInternal: isRequireAuthentication,
    AuthenticatingInternal: Authenticating,
  })
);

export default OidcSecure;
