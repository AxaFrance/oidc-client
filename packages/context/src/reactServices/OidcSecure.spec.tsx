import React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { useOidcSecure, withOidcSecureWithInjectedFunctions, OidcSecureWithInjectedFunctions } from './OidcSecure';
import { AuthenticationContext } from '../oidcContext';
import { oidcContext } from '../oidcContext/AuthenticationContext';

describe('OidcSecure tests suite', () => {
  const values: oidcContext = {
    isEnabled: true,
    oidcUser: { user: { name: 'tom', isExpired: false } },
    authenticating: () => (
      <div>
        <span>Authenticating...</span>
      </div>
    ),
  };

  const getWrapper = (vals = values) => ({ children }) => (
    <AuthenticationContext.Provider value={vals}>{children}</AuthenticationContext.Provider>
  );

  const userManager = 'user Manager mock';
  const getUserManagerInternal = () => userManager;
  const authenticateUserInternal = jest.fn(() => jest.fn());
  const oidcLogInternal = {
    info: jest.fn(),
    debug: jest.fn(),
    error: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });
  const AuthenticatingInternal = () => (
    <div>
      <span>Default authenticating page</span>
    </div>
  );
  const isRequireAuthenticationInternal = user => !user || (user && user.expired === true);
  const WrappedComponent = () => (
    <div>
      <span>Testing page</span>
    </div>
  );
  const mockedProps = {
    location: '/myurl',
    history: { push: () => {} },
    authenticateUserInternal,
    getUserManagerInternal,
    isRequireAuthenticationInternal,
    AuthenticatingInternal,
  };
  const history = { push: () => {} };
  const TestingComponent = ({ props1 }) => <div data-testid="componentMount">{props1}</div>;

  it('should call authentication when enabled, and display page if user connected', () => {
    const { result } = renderHook(
      () =>
        useOidcSecure(
          authenticateUserInternal,
          userManager,
          '/locationUser',
          history,
          oidcLogInternal,
          AuthenticatingInternal,
          isRequireAuthenticationInternal,
          WrappedComponent
        ),
      {
        wrapper: getWrapper(),
      }
    );

    expect(authenticateUserInternal).toHaveBeenCalledWith(userManager, '/locationUser', history);
    expect(result.current()).toEqual(
      <div>
        <span>Testing page</span>
      </div>
    );
  });

  it('should display authenticating component when user is not connected', () => {
    const withoutUserValues = { ...values, oidcUser: undefined };

    const { result } = renderHook(
      () =>
        useOidcSecure(
          authenticateUserInternal,
          userManager,
          '/locationUser',
          history,
          oidcLogInternal,
          AuthenticatingInternal,
          isRequireAuthenticationInternal,
          WrappedComponent
        ),
      {
        wrapper: getWrapper(withoutUserValues),
      }
    );

    expect(result.current()).toEqual(
      <div>
        <span>Authenticating...</span>
      </div>
    );
  });

  it('should display default authenticating component when user is not connected and ahenticating page is not set', () => {
    const withoutUserValues = { ...values, oidcUser: undefined, authenticating: undefined };

    const { result } = renderHook(
      () =>
        useOidcSecure(
          authenticateUserInternal,
          userManager,
          '/locationUser',
          history,
          oidcLogInternal,
          AuthenticatingInternal,
          isRequireAuthenticationInternal,
          WrappedComponent
        ),
      {
        wrapper: getWrapper(withoutUserValues),
      }
    );
    expect(result.current()).toEqual(
      <div>
        <span>Default authenticating page</span>
      </div>
    );
  });

  it('should NOT call authentication when disabled', () => {
    const { result } = renderHook(
      () =>
        useOidcSecure(
          authenticateUserInternal,
          userManager,
          '/locationUser',
          history,
          oidcLogInternal,
          AuthenticatingInternal,
          isRequireAuthenticationInternal,
          WrappedComponent
        ),
      {
        wrapper: getWrapper({ ...values, isEnabled: false }),
      }
    );
    expect(authenticateUserInternal).not.toHaveBeenCalled();
    expect(result.current()).toEqual(
      <div>
        <span>Testing page</span>
      </div>
    );
  });

  it('OidcSecureWithInjectedFunctions should call authentication and AuthenticatingComponent during authentication process', () => {
    const contextMock = {
      ...values,
      isEnabled: true,
      oidcUser: null,
      authenticating: undefined,
    };

    const { queryByTestId, asFragment } = render(
      <OidcSecureWithInjectedFunctions {...mockedProps}>
        <TestingComponent />
      </OidcSecureWithInjectedFunctions>,
      {
        wrapper: getWrapper({ ...contextMock }),
      }
    );
    expect(asFragment()).toMatchSnapshot();
    expect(queryByTestId('componentMount')).toBeNull();
  });

  it('OidcSecureWithInjectedFunctions should show component if user connected', () => {
    const contextMock = {
      ...values,
      isEnabled: true,
      oidcUser: { user: 'tom' },
    };

    const { queryByTestId, asFragment } = render(
      <OidcSecureWithInjectedFunctions {...mockedProps}>
        <TestingComponent />
      </OidcSecureWithInjectedFunctions>,
      {
        wrapper: getWrapper({ ...contextMock }),
      }
    );
    expect(asFragment()).toMatchSnapshot();
    expect(queryByTestId('componentMount')).not.toBeNull();
  });

  it('withOidcSecureWithInjectedFunctions should call authentication and AuthenticatingComponent during authentication process', () => {
    const contextMock = {
      ...values,
      isEnabled: true,
      oidcUser: null,
      authenticating: undefined,
    };

    const HocComponent = withOidcSecureWithInjectedFunctions(TestingComponent);
    const { asFragment } = render(<HocComponent {...mockedProps} props1="ValSupp" />, {
      wrapper: getWrapper({ ...contextMock }),
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('withOidcSecureWithInjectedFunctions should show component if user connected', () => {
    const contextMock = {
      ...values,
      isEnabled: true,
      oidcUser: { user: 'tom' },
    };
    const HocComponent = withOidcSecureWithInjectedFunctions(TestingComponent);

    const { queryByTestId, asFragment } = render(<HocComponent {...mockedProps} props1="ValSupp" />, {
      wrapper: getWrapper({ ...contextMock }),
    });
    expect(asFragment()).toMatchSnapshot();
    expect(queryByTestId('componentMount')).not.toBeNull();
  });
});
