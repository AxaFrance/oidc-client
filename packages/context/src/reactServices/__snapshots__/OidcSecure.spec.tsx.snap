// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OidcSecure tests suite OidcSecureWithInjectedFunctions should call authentication and AuthenticatingComponent during authentication process 1`] = `
<DocumentFragment>
  <div>
    <span>
      Default authenticating page
    </span>
  </div>
</DocumentFragment>
`;

exports[`OidcSecure tests suite OidcSecureWithInjectedFunctions should show component if user connected 1`] = `
<DocumentFragment>
  <div
    data-testid="componentMount"
  />
</DocumentFragment>
`;

exports[`OidcSecure tests suite withOidcSecureWithInjectedFunctions should call authentication and AuthenticatingComponent during authentication process 1`] = `
<DocumentFragment>
  <div>
    <span>
      Default authenticating page
    </span>
  </div>
</DocumentFragment>
`;

exports[`OidcSecure tests suite withOidcSecureWithInjectedFunctions should show component if user connected 1`] = `
<DocumentFragment>
  <div
    data-testid="componentMount"
  >
    ValSupp
  </div>
</DocumentFragment>
`;
